pmd-java/src/main/java/net/sourceforge/pmd/lang/java/oom/ClassStats.java

```diff
@@ -159,14 +159,14 @@
      *
      * @return The result of the computation, or {@code Double.NaN} if it couldn't be performed
      */
-    /* default */ double computeWithResultOption(OperationMetricKey key, ASTAnyTypeDeclaration node,
+    /* default */ double computeWithResultOption(MetricKey<OperationMetric> key, ASTAnyTypeDeclaration node,
                                                  boolean force, MetricVersion version, ResultOption option) {
 
         List<ASTMethodOrConstructorDeclaration> ops = findOperations(node, false);
 
         List<Double> values = new ArrayList<>();
         for (ASTMethodOrConstructorDeclaration op : ops) {
-            if (key.supports(op)) {
+            if (key.getCalculator().supports(op)) {
```

Comment:
I think inheritance is fine to express the differences between them. That allows us to have one type for class metrics and another for operation metrics, which can be used to put compile time constraints on what metric can be calculated on what node. For example, we can forbid calculating a class metric on an operation node (see the `get` overloads in `Metrics`).

The `supports` method aims at even finer capability checking. For instance some class metrics may not be calculated on interfaces, even though the compile time check above would allow it (as interfaces, classes, enums and annotation type declarations are grouped under `ASTAnyTypeDeclaration`). You can see the defaults of this method in `AbstractMetric`.

The fact that `supports` had an `AccessNode` parameter was just poor design. I did that so that metrics, whether class or operation metrics, could all extend a single `AbstractMetric` class. To do that, `supports` had to be declared in the `Metric` interface, which is why its argument's type could not be too specific. But ideally the interfaces shouldn't even hint that a class metric could support any `AccessNode`, so it would be good to change that.